internal/services/extendedlocation/extended_location_custom_location.go

```diff
@@ -0,0 +1,327 @@
+				}
+				if props.HostResourceId != nil {
+					state.HostResourceId = pointer.From(props.HostResourceId)
+				}
+				if props.HostType != nil {
+					state.HostType = string(pointer.From(props.HostType))
+				}
+				if props.Namespace != nil {
+					state.Namespace = pointer.From(props.Namespace)
+				}
+
+				if props.Authentication != nil && props.Authentication.Type != nil && props.Authentication.Value != nil {
+					state.Authentication = []AuthModel{
+						{
+							Type:  pointer.From(props.Authentication.Type),
+							Value: pointer.From(props.Authentication.Value),
+						},
+					}
+				}
+
+				return metadata.Encode(&state)
+			}
+			return nil
+		},
+	}
+}
+
+func (r CustomLocationResource) Delete() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.ExtendedLocation.CustomLocationsClient
+			id, err := customlocations.ParseCustomLocationID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			metadata.Logger.Infof("deleting %s", *id)
+
+			if resp, err := client.Delete(ctx, *id); err != nil {
+				if !response.WasNotFound(resp.HttpResponse) {
+					return fmt.Errorf("deleting %s: %+v", *id, err)
+				}
+			}
+
+			return nil
+		},
+	}
+}
+
+func (r CustomLocationResource) Update() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.ExtendedLocation.CustomLocationsClient
+			id, err := customlocations.ParseCustomLocationID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			var state CustomLocationResourceModel
+			if err := metadata.Decode(&state); err != nil {
+				return fmt.Errorf("decoding %+v", err)
+			}
+
+			customLocationProps := customlocations.CustomLocationProperties{}
+			d := metadata.ResourceData
+
+			if d.HasChanges("authentication") {
+				if len(state.Authentication) > 0 {
+					auth := state.Authentication[0]
+					customLocationProps.Authentication = &customlocations.CustomLocationPropertiesAuthentication{
+						Type:  pointer.To(auth.Type),
+						Value: pointer.To(auth.Value),
+					}
+				}
+			}
+
+			if d.HasChange("cluster_extension_ids") {
+				customLocationProps.ClusterExtensionIds = pointer.To(state.ClusterExtensionIds)
+			}
+
+			if d.HasChange("display_name") {
+				customLocationProps.DisplayName = pointer.To(state.DisplayName)
+			}
+
+			if d.HasChange("host_resource_id") {
+				customLocationProps.HostResourceId = pointer.To(state.HostResourceId)
+			}
+
+			if d.HasChange("host_type") {
+				hostType := customlocations.HostType(state.HostType)
+				customLocationProps.HostType = pointer.To(hostType)
+			}
+
+			props := customlocations.PatchableCustomLocations{
+				Properties: pointer.To(customLocationProps),
+			}
+
+			if _, err := client.Update(ctx, *id, props); err != nil {
```

Comment:
Behaviourally, `Update` uses a PATCH, which can prevent removal of values in some cases, do we have sufficient test coverage to ensure this method is appropriate for use in the provider and exhibits the correct behaviour for users removing properties from their configuration? Would `CreateOrUpdate` be more appropriate here?

Reply:
I'm not sure if I follow. For this resource, there aren't many attributes updatable. For example, `authentication` block, as communicated on another thread, cannot be covered by acc test.

Reply:
The number of properties doesn't really matter, it's the behaviour of the API and compatibility with Terraform's expectations. So, if we have one or more properties that are allowed to be removed entirely, such as `authentication`, we need to know how to unset those. If sending a `nil` for a value doesn't remove/unset it (as per the HTTP spec, this should be a no-op) we need to send the appropriate value to remove it.

To the point of `Update` vs `CreateOrUpdate`, the "location is required" error will not happen if we use a `Get` to retrieve the existing data, then simply update that with any user changes before sending it back, rather than using the `Patch` only method?

Reply:
Make sense. I've changed this to `CreateOrUpdate`.